internals-js/src/federation.ts

```diff
@@ -1116,34 +1140,23 @@ export function parseFieldSetArgument({
       throw e;
     }
 
-    const nodes = sourceASTs(directive);
-    if (e.nodes) {
-      nodes.push(...e.nodes);
-    }
-    let msg = e.message.trim();
-    // The rule for validating @requires in fed 1 was not properly recursive, so people upgrading
-    // may have a @require that selects some fields but without declaring those fields on the
-    // subgraph. As we fixed the validation, this will now fail, but we try here to provide some
-    // hint for those users for how to fix the problem.
-    // Note that this is a tad fragile to rely on the error message like that, but worth case, a
-    // future change make us not show the hint and that's not the end of the world.
-    if (msg.startsWith('Cannot query field')) {
-      if (msg.endsWith('.')) {
-        msg = msg.slice(0, msg.length - 1);
-      }
-      if (directive.name === keyDirectiveSpec.name) {
-        msg = msg + ' (the field should be either be added to this subgraph or, if it should not be resolved by this subgraph, you need to add it to this subgraph with @external).';
-      } else {
-        msg = msg + ' (if the field is defined in another subgraph, you need to add it to this subgraph with @external).';
-      }
-    }
-
-    const codeDef = errorCodeDef(e) ?? ERROR_CATEGORIES.DIRECTIVE_INVALID_FIELDS.get(directive.name);
-    throw codeDef.err({
-      message: `${fieldSetErrorDescriptor(directive)}: ${msg}`,
-      nodes,
-      originalError: e,
-    });
+    throw handleFieldSetValidationError(
+      directive,
+      e,
+      (msg: string) => {
+        if (msg.startsWith('Cannot query field')) {
+          if (msg.endsWith('.')) {
+            msg = msg.slice(0, msg.length - 1);
+          }
+          if (directive.name === keyDirectiveSpec.name) {
+            msg = msg + ' (the field should be either be added to this subgraph or, if it should not be resolved by this subgraph, you need to add it to this subgraph with @external).';
```

Comment:
**nitpick** I know this was already there, but since I see it

```suggestion
            msg = msg + ' (the field should either be added to this subgraph or, if it should not be resolved by this subgraph, you need to add it to this subgraph with @external).';
```